benchmark/C/Savina/src/concurrency/Dictionary.lf

```diff
@@ -7,90 +9,25 @@
  * to query the LinkList.
+ *
+ * @author Thee Ho
+ * @author Matt Chorlian
  */
 
-/* [[[cog
-# This file is a code generator using the python module cog:
-# See https://nedbatchelder.com/code/cog/
-#
-# All instructions for code  generation are in-lined in comments
-# like this one. With that you can use this file as a normal source file
-# but also to generate code.
-# 
-# To change the generated code in-line within this file run:
-# $ python -m cog -r this-file.lf
-# To generate a new file from this file stripping the generator code in the process run:
-# $ python -m cog -d -o output-file.lf this-file.lf
-#
-# Use the command line option -D to specify generator parameters, for example:
-# $ python -m cog -r -D parameter=100 this-file.lf
-#
-# Generator parameters used in this file:
-# -D num_workers=20
-# 
-]]] */
-// [[[end]]]
-
-/* [[[cog
-  # force existence, type and default values of generator parameters
-  if 'num_workers' in globals():
-    num_workers = int(num_workers)
-  else:
-    globals()['num_workers'] = 20
-  
-  # output the current value of the generator parameters used in the last generation run
-  cog.outl(f'// Generated file with the following parameters:')
-  cog.outl(f'// num_workers = {num_workers}')
-]]] */
-// Generated file with the following parameters:
-// num_workers = 20
-// [[[end]]] 
-
-/* [[[cog
-# This file is a code generator using the python module cog:
-# See https://nedbatchelder.com/code/cog/
-#
-# All instructions for code  generation are in-lined in comments
-# like this one. With that you can use this file as a normal source file
-# but also to generate code.
-# 
-# To change the generated code in-line within this file run:
-# $ python -m cog -r this-file.lf
-# To generate a new file from this file stripping the generator code in the process run:
-# $ python -m cog -d -o output-file.lf this-file.lf
-#
-# Use the command line option -D to specify generator parameters, for example:
-# $ python -m cog -r -D parameter=100 this-file.lf
-#
-# Generator parameters used in this file:
-# -D num_workers=20
-# 
-
-]]] */
-// [[[end]]]
-
-/* [[[cog
-  # force existence, type and default values of generator parameters
-  if 'num_workers' in globals():
-    num_workers = int(num_workers)
-  else:
-    globals()['num_workers'] = 20
-  
-  # output the current value of the generator parameters used in the last generation run
-  cog.outl(f'// Generated file with the following parameters:')
-  cog.outl(f'// num_workers = {num_workers}')
-]]] */
-// Generated file with the following parameters:
-// num_workers = 20
-// [[[end]]] 
 
 target C {
-    fast: true,
-    threads: 1,
-    tracing: true
+    /* [[[cog
+      if (threaded_runtime=="True"):
+          cog.outl(f"threads: {threads}")
+      else:
+          cog.outl("threads: 0")
+    ]]] */
+    threads: 0
+    /// [[[end]]]
 };
 
-reactor Dictionary(num_workers:int(20)){
+
+reactor DictionaryImpl(num_workers:int(20)){
     preamble {=
 		struct node{
```

Comment:
The main problem with this benchmark is the implementation of the dictionary structure. I don't think it is complete and realistic, but I don't know if we want to invest in implementing a complete dictionary in C.... Maybe there is an implementation with a permissive license somewhere in the internet?

I cannot comment on the code below unfortunately, so I will summarize some of my thoughts here:

- In the preamble (definition of the dictionary): 
  - `table` does not seem to be initialized. There is no malloc for it
  - Elements can be inserted, but there is no function for cleaning up the dictionary at shutdown
  - The hash function seems to be overly simplistic
- In the Dictionary reactor:
  - The initialization in startup is strange, the dictionary is supposed to be empty
  - The C++ version does not distinguish read and write ports. It uses a single request port. This should be a bit faster as there is only a single reaction in the dictionary that needs to be processed.
  - The C++ version introduces a logical action in the Dictionary to break up the dependency cycle. In this implementaton, the logical action is contained in the workers. While both options should work fine, the one implemented for C++ is more efficient, as per iteration it has only a single call to schedule. In this version, each worker schedules its logical action resulting in multiple calls and multiple events being created.
- In the Worker reactor:
  - The Savina benchmark uses the Scala random number generator instead of the PseudoRandom generator. This is a bad design in Savina. However, we need to think how to best translate this to C. Simply using `rand()` does not work, since each worker needs its own generator initiailized with its own seed. In C++, we use [`std::minstd_rand`](http://www.cplusplus.com/reference/random/minstd_rand/) to create an instance of a random number generator. Is there something similar for C?
  - The workers limit the produced Dictionary keys to be smaller than 1000. I think this is due to the limited size of the dictionary implementation. However, we should be able to deal with arbitrary keys.